src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1628,6 +1628,12 @@ private void ReportResultsToSubmission(BuildResult result)
 
                     _overallBuildSuccess = _overallBuildSuccess && (_buildSubmissions[result.SubmissionId].BuildResult.OverallResult == BuildResultCode.Success);
 
+                    if (submission.BuildRequestData != null && submission.BuildRequestData.Flags.HasFlag(BuildRequestDataFlags.ClearProjectRootElementCacheAfterBuild))
+                    {
+                        // Reset the project root element cache if specified which ensures that projects will be re-loaded from disk
+                        _buildParameters?.ProjectRootElementCache?.Clear();
```

Comment:
Out of proc nodes have their own copy of the project root element cache. So resetting just the main node will not clear out the nodes, which would still lead to the stale cache issue :(

Either the build manager needs to shut down all nodes, or you need to implement another packet that tells the nodes to reset their caches.

Reply:
Apparently out of proc nodes setup the cache to [reload from disk](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Node/OutOfProcNode.cs#L180). So if restore changes any existing imports, the nodes should detect the change and replace the cache entries with updated values.